src/track/track.cpp

```diff
@@ -160,50 +164,60 @@ void Track::importMetadata(
+        // knowing the sample rate. Bug #1020438.
+        // Only use the imported BPM if the current beat grid is not valid!
+        // Reason: The BPM value in the metadata might be  imprecise (normalized
+        // or rounded), e.g. ID3v2 only supports integer values.
+        const bool beatsAndBpmModified = importedBpm.hasValue() &&
+                trySetBpmWhileLocked(importedBpm.getValue(), false);
+        modified |= beatsAndBpmModified;
+        const auto newBpm = getBpmWhileLocked();
+
+        // Only update the current key with a valid value. Otherwise preserve
+        // the existing value.
+        const bool keysModified = !importedKeyText.isEmpty() &&
+                KeyUtils::guessKeyFromText(importedKeyText) != mixxx::track::io::key::INVALID &&
+                m_record.updateGlobalKeyText(importedKeyText, mixxx::track::io::key::FILE_METADATA);
+        modified |= keysModified;
+        const auto newKey = m_record.getGlobalKey();
+
+        // explicitly unlock before emitting signals
+        markDirtyAndUnlock(&lock, modified);
+
+        if (beatsAndBpmModified) {
+            emitBeatsAndBpmUpdated(newBpm);
         }
-
-        // FIXME: Move the Track::setCuePoints call to another location,
-        // because we need the sample rate to calculate sample
-        // positions for cues (and *correct* sample rate isn't known here).
-        //
-        // If the Serato tags are empty they are not present. In this case
-        // all existing metadata must be preserved!
-        if (!newSeratoTags.isEmpty()) {
-            tryImportBeats(newSeratoTags.importBeats(), newSeratoTags.isBpmLocked());
-            importCueInfos(newSeratoTags.importCueInfos());
-            setColor(newSeratoTags.getTrackColor());
+        if (keysModified) {
+            emitKeysUpdated(newKey);
+        }
+        if (oldReplayGain != newReplayGain) {
+            emit replayGainUpdated(newReplayGain);
         }
-
-        // implicitly unlocked when leaving scope
-    }
-
-    // Need to set BPM after sample rate since beat grid creation depends on
-    // knowing the sample rate. Bug #1020438.
-
-    // Only use the imported BPM if the beat grid is not valid!
-    // Reason: The BPM value in the metadata might be normalized
-    // or rounded, e.g. ID3v2 only supports integer values.
-    if (!m_pBeats || !mixxx::Bpm::isValidValue(m_pBeats->getBpm())) {
-        trySetBpm(newBpm.getValue());
     }
 
-    if (!newKey.isEmpty()
-            && KeyUtils::guessKeyFromText(newKey) != mixxx::track::io::key::INVALID) {
-        setKeyText(newKey, mixxx::track::io::key::FILE_METADATA);
+    // FIXME: Move the Track::setCuePoints call to another location,
+    // because we need the sample rate to calculate sample
+    // positions for cues (and *correct* sample rate isn't known here).
+    //
+    // If the Serato tags are empty they are not present. In this case
+    // all existing metadata must be preserved. Importing Serato metadata
```

Comment:
@Holzhaus Even the comment explains why Serato tags are imported **after** all other metadata.

Reply:
That's my I'm asking. With this PR reimporting does not overwrite the beatgrid (as shown in the Screencast). I had to reset it first.

Reply:
I will undo some changes and remove the `overwriteValidBeats` parameter. It is not needed and it looks like I messed up the condition.

Reply:
Hm, no, the condition seemed to be correct. Anyway, I have removed the intermediate changes  and force pushed. The boolean parameter was an ugly workaround, not needed anymore.

Reply:
In the old code (before this PR) sometimes the `Track` object was still locked while importing the Serato tags and sometimes not. Now the behavior should be consistent, i.e. always unlocked before finally importing the Serato tags.